packages/flutter/lib/src/widgets/restoration_properties.dart

```diff
@@ -0,0 +1,250 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'editable_text.dart';
+import 'restoration.dart';
+
+/// A [RestorableProperty] that makes the wrapped value accessible to the owning
+/// [State] object via the [value] getter and setter.
+///
+/// Whenever a new [value] is set, [didUpdateValue] is called. Subclasses should
+/// call [notifyListeners] from this method if the new value changes what
+/// [toPrimitives] returns.
+///
+/// See also:
+///
+///  * [RestorableProperty], which is the super class of this class.
+///  * [RestorationMixin], to which a [RestorableValue] needs to be registered
+///    in order to work.
+///  * [RestorationManager], which provides an overview of how state restoration
+///    works in Flutter.
+abstract class RestorableValue<T> extends RestorableProperty<T> {
```

Comment:
Instead of:


```dart
  RestorableNum<int> _counter = RestorableNum<int>(0);
  RestorableTextEditingController _textController = 
      RestorableTextEditingController();

  @override
  void restoreState() {
    registerForRestoration(_counter, RestorationId('counter'));
    registerForRestoration(_textController, RestorationId('name'));
  }
```

Would it make sense to have:


```dart
  RestorableNum<int> _counter = RestorableNum<int>(0, id: 'counter');
  RestorableTextEditingController _textController = 
      RestorableTextEditingController(id: 'name');

  @override
  void restoreState() {
    registerForRestoration(_counter);
    registerForRestoration(_textController);
  }
```

Reply:
That's possible. But I figured that keeping the line with the instantiation free of the ID clutter makes the code a little easier to read. In your IDE when you click through to go to the definition of the `_counter ` variable you're most likely not interested in its restorationId. Furthermore, it keeps all IDs neatly in one place (the restoreState method) making it easier to see with one glance what IDs are in use.

Reply:
Why would you need to return the RestorationID? Either you return an un-registered property and whoever receives the property is in charge of registering it (under an ID of their choosing to ensure it doesn't clash with anything else in their namespace) or you return a registered property in which cause whoever receives the property doesn't need to care about its ID.

Reply:
I had state-management libraries in mind – which will have to implement a way to interact with the restoration API outside of a `State`/the widget tree.

Provider will definitely want to expose a way for users to define restorable properties, where they would define both the ID and the property from outside a `State`.

It may not be too bad, as we can use `Map<Id, Property>` instead of a tuple. I need to play around it.

Reply:
The code registering the property (i.e. the code with the bucket) must be the one to come up with the IDs, otherwise it can't properly avoid name clashes.